number_theory/numeric.h

```diff
@@ -92,10 +95,59 @@ auto pow(T base, U exponent) -> decltype(std::pow(base, exponent)) {
   return std::pow(base, exponent);
 }
 
+// Returns the |n|-th root of |x| rounded to integer toward zero.
+// The argument |n| shoud be positive.
+// Throws domain_error exception if the root does not exist.
+// The time complexity is O(log |x|).
+//
+// This overload is only enabled for x < 2^64, because it uses uint64_t
+// to avoid overflows during the computation.
+template <typename T,
+          typename U,
+          std::enable_if_t<std::numeric_limits<T>::digits <= 64, bool> = true>
+T iroot(const T &x, const U &n) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && std::numeric_limits<U>::is_integer,
+      "iroot arguments must be integers.");
+  if (n < 0)
+    throw std::invalid_argument("iroot argument |n| should be positive.");
+  if (n == 0)
+    throw std::domain_error("The 0-th root does not exist.");
+  if (x < 0 && n % 2 == 0)
+    throw std::domain_error("Negative numbers do not have even roots.");
+
+  if (n == 1)
+    return x;
+
+  // For each n, the table stores the maximum y that y^n does not overflow.
+  static const int max_size = 64;
+  static const uint64_t y_limits[max_size] = {
+      0,  0,  4294967295, 2642245, 65535, 7131, 1625, 565, 255, 138, 84, 56, 40,
+      30, 23, 19,         15,      13,    11,   10,   9,   8,   7,   6,  6,  5,
+      5,  5,  4,          4,       4,     4,    3,    3,   3,   3,   3,  3,  3,
+      3,  3,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2,  2,
+      2,  2,  2,          2,       2,     2,    2,    2,   2,   2,   2,  2};
+  uint64_t y_limit = (n >= max_size ? 1 : y_limits[n]);
+
+  // Binary search for the maximum y that y^n <= abs(x).
+  uint64_t abs_x = unsigned_abs(x);
+  uint64_t y_min = 0;
```

Comment:
This binary search implementation is not so common. I suggest implementing in the most common way such as:
```cpp
uint64_t left = 0;
uint64_t right = std::min(abs_x, y_limit);
while (left <= right) {
  uint64_t mid = std::mid_point(left, right);
  if (pow(mid, n) <= abs_x)
    left = mid + 1;
  else
    right = mid - 1;
}
return sign(x) * static_cast<T>(left - 1);
```
Usually, people even introduce another variable and update it in the conditional branch which is taken when a valid answer is found which is the case where I updated left above. That way, people avoid returning `left - 1` or `right + 1` in the end. However, I was worried about cost of copying there. Maybe, the cost is negligible.

Another thing I just remembered is that Google's C++ style guide says that if a conditional statement has `else` branch, it should always use curly braces. I don't know if we agree with them or not. I like omitting curlies when the bodies are simple even with multiple cases. Still, it might be good to reconsider our decisions while we are working on this.

Reply:
There are many ways to implement binary search, and I don't think one way is more common than another. Actually the version you provided does not look better, because it has many `+1` and `-1` which confuse me.

Reply:
I am looking forward to switch to `std::ranges::binary_search` once this c++20 feature is supported by clang-12.

Reply:
> There are many ways to implement binary search, and I don't think one way is more common than another. Actually the version you provided does not look better, because it has many `+1` and `-1` which confuse me.

It is interesting that you are not used to this version. I thought that was the most common one since that's what I saw in almost any tutorial and book. I was using your version during codeforces competitions but then I realized most editorials had the version I posted above, including red coders or even LGMs. I am not sure about it either, but even Errichto seems to prefer that version:
https://www.youtube.com/watch?v=GU7DpgHINWQ

> I am looking forward to switch to `std::ranges::binary_search` once this c++20 feature is supported by clang-12.

This sounds good, but I would also consider `std::ranges::lower_bound` since we are not searching for a specific value.

Reply:
I submitted a commit for a new style, although it looked more confusing to me. You can decide whether it is better and which way we go for.